### PR TITLE
Generate bundle.mlir for symbolic batch dimension for pointwise ops.

### PR DESCRIPTION
**Pull request by nethra-khandige**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:
Emits symbolic `input_arg` parameters with `granularity` and `max_value` in `bundle.mlir` for dimensions marked dynamic via mark_dynamic.

`bundle.mlir` output changes (`bundle.py`):

- Dimension symbols now appear as typed function parameters: `!sdscbundle.input_arg<index, granularity=G, max_value=M>`, independent of `use_symbols`
- Corresponding `input_arg_extract` ops unpack each parameter into a plain index SSA value
- Dimension symbols with the same pytorch_sym across multiple SDSCs share a single parameter (deduplicated by pytorch_sym)
- `sdsc_execute` always emits `symbol_ids`=[...]; dimension symbol IDs (from `dimToSymbolMapping_`) appear first, followed by HBM address IDs

Symbol extraction (`bundle.py`):

- `_extract_symbol_ids` now scans `dimToSymbolMapping_` before scheduleTree_ to collect dimension IDs ahead of HBM address IDs, matching the operand order in sdsc_execute

Tests (`tests/inductor/test_symbolic_dim_bundle.py`):

- New test file covering `_extract_symbol_ids` and `generate_bundle` for the dimension-symbol path
- existing `tests/dynamic_shapes/mark_dynamic.py`:modified to handle failure as expected for now.

Additional logs/bundle.mlir:
[sym_bundle.zip](https://github.com/user-attachments/files/30148550/sym_bundle.zip)

#### Which issue(s) this PR is related to:
https://github.com/torch-spyre/torch-spyre/issues/43, https://github.com/torch-spyre/torch-spyre/issues/2500

#### Does this PR introduce a user-facing change?
Yes. Users can now use mark_dynamic(dim, min=..., max=...) on a batch dim for pointwise ops to generate bundle.mlir along with sdsc.json

### CODE REVIEW

**Comment by nethra-khandige** [comment]
This PR enables bundle.mlir generation for symbolic batch size dimension with max_size and granularity. Example Test file:
```
import torch
import torch._dynamo as dynamo
import torch.nn.functional as F

DEVICE = torch.device("spyre")
torch.manual_seed(0xAFFE)

def gelu_fn(a):
    return F.gelu(a)

x = torch.rand(560, 1024, dtype=torch.float16)

# Mark dim 0 as dynamic with an explicit upper bound.
x_device = x.to(DEVICE)
dynamo.mark_dynamic(x_device, 0, min=56, max=616)
compiled_fn = torch.compile(gelu_fn)

compiled_result = compiled_fn(x_device).cpu()
print(f"Spyre Compiled result\n{compiled_result}")
```


Bundle.mlir (generated as per provided by the DeepTools team) and detailed logs
[sym_bundle.zip](https://github.com/user-attachments/files/30175494/sym_bundle.zip)

File: tests/dynamic_shapes/mark_dynamic.py [hunk 1]
```diff
@@ -42,10 +42,17 @@ def gelu_fn(a):
...
-cpu_delta = torch.abs(compiled_result - cpu_result).max()
-print(f"Max delta Compiled Spyre vs. CPU: {cpu_delta}")
+# EXPECTED FAILURE: dispatch currently fails in deeptools
+# (tracked in #2434).
+# Kept as a reproducer for that issue rather than removed or silently
+
+try:
+    compiled_result = compiled_fn(x_device).cpu()
+except Exception as e:
+    print(f"Expected dispatch failure (see #2434): {e}")
```
Comment (Vikram123-byte) [inline tests/dynamic_shapes/mark_dynamic.py]: Do not treat every compiled execution exception as an expected dispatch failure.

This catches failures from compilation, lowering, invalid inputs, and unrelated runtime bugs, allowing the test to pass silently. Catch the concrete exception (or validate the known #2434 failure signature) and re-raise anything else.

File: torch_spyre/_inductor/codegen/bundle.py [hunk 2]
```diff
@@ -193,6 +238,15 @@ def generate_bundle(
...
                 )
+            for sym_idx in dimension_sym_indices:
+                dim_sk = symbol_kinds[sym_idx]
+                name = dim_param_names[sym_idx]
+                f.write(
+                    f"\t\t{name} = sdscbundle.input_arg_extract value from"
+                    f" {name}_base : !sdscbundle.input_arg<index, "
+                    f"granularity={dim_sk.granularity}, max_value={dim_sk.max_value}>"
+                    " -> index\n"
+                )
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/codegen/bundle.py:258]: Extract the dimension input_arg type string into a helper to avoid drift.

The !sdscbundle.input_arg<index, granularity=..., max_value=...> type string is built independently in the param declaration (Lines 225-228) and the extract op (Lines 244-248). If one is edited without the other, the two would silently diverge and produce a generated bundle.mlir where declared and extracted types don't match.

File: tests/dynamic_shapes/mark_dynamic.py (hunk 1 above)
Comment (nethra-khandige) [reply tests/dynamic_shapes/mark_dynamic.py]: Thanks. Changed to catch the expected exception message.

File: torch_spyre/_inductor/codegen/bundle.py (hunk 2 above)
Comment (nethra-khandige) [reply torch_spyre/_inductor/codegen/bundle.py:258]: Thank you for the suggestion. Implemented a helper to ensure there is no drift.

File: torch_spyre/_inductor/codegen/bundle.py [hunk 3]
```diff
@@ -123,11 +127,20 @@ def generate_bundle(
...
+    for sdsc_idx, (_, _, _, local_kinds) in enumerate(compiled):
+        local_dim_ordinal = 0
+        for lk in local_kinds:
+            if lk.is_dimension:
+                local_dim_ordinal += 1
+                sym_idx_to_dim_origin[len(symbol_kinds)] = (
+                    sdsc_idx,
+                    local_dim_ordinal,
+                )
+            symbol_kinds.append(lk)
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/codegen/bundle.py:155]: Keep address-symbol lowering gated by use_symbols.

Unconditionally retaining SymbolKinds makes the later declaration loop process pool, kernel_slice, and kernel_derived symbols even when use_symbols=False. Those branches emit %pool or %arg_*, but Lines 216-239 do not declare/extract those values in that mode, producing invalid MLIR for static-address bundles.

**Review by Vikram123-byte — COMMENTED** [review]
Left one minor comment @nethra-khandige 
Rest looks good

File: torch_spyre/_inductor/codegen/bundle.py (hunk 3 above)
Comment (nethra-khandige) [reply torch_spyre/_inductor/codegen/bundle.py:155]: Thank you. Fixed it with an assertion in the flattening loop that fails if a non-dimension SymbolKind ever shows up when use_symbols=False, since that invariant was previously only enforced implicitly in compute_ops.py.

File: torch_spyre/_inductor/codegen/bundle.py [hunk 4]
```diff
@@ -140,6 +140,11 @@ def generate_bundle(
                     sdsc_idx,
                     local_dim_ordinal,
                 )
+            elif not use_symbols:
+                raise AssertionError(
+                    "use_symbols=False but compute_op_spec registered a "
+                    f"non-dimension SymbolKind ({lk.kind!r}); address-symbol."
+                )
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/codegen/bundle.py:154]: Avoid leaving partial bundle artifacts on validation failure.

_compile_specs runs before this check and writes sdsc_N.json files, but this exception occurs before bundle.mlir is generated. If output_dir is reused or inspected after the failure, it can contain a mixed or incomplete bundle. Generate into a staging directory and publish atomically, or remove all artifacts when validation fails.

**Review by dgrove-oss — COMMENTED** [review]
Looks good to me. Holding off on approval until @Vikram123-byte comment is resolved.

Code review below:

-----

PR #3238: Generate bundle.mlir for symbolic batch dimension for pointwise ops

Summary

Extends bundle.py to emit mark_dynamic dimension symbols as typed !sdscbundle.input_arg<index, granularity=G, max_value=M> function parameters in bundle.mlir, independent of use_symbols, and reorders _extract_symbol_ids to collect dimension IDs before HBM address IDs to match the new operand order. Adds a dedicated unit test file and flips a mark_dynamic.py E2E reproducer into an expected-failure-with-specific-message test tracked against issue #2434.

Review

BLOCKERs

None found.

SUGGESTIONs

- [ ] A few added lines in tests/inductor/test_symbolic_dim_bundle.py exceed the 88-char limit (e.g. the two input_arg<...> assertion strings and one comment). Won't fail CI since ruff's default ruleset doesn't enforce E501, but worth a pass with the formatter.
- [ ] The new AssertionError guard in generate_bundle ("use_symbols=False but compute_op_spec registered a non-dimension SymbolKind") is a good defensive invariant, but it's currently unreachable dead code from the caller's perspective unless compute_ops.py's contract changes — a one-line comment noting why it's safe (i.e. local_symbol_kind is only ever populated when use_symbols=True, per generate_sdsc) would help future readers trust it rather than wonder if it's a latent bug.

QUESTIONs

- [ ] mark_dynamic.py's dispatch-failure path is explicitly tied to issue #2434 ("runtime isn't yet wired to supply that extra argument"). Is there a rough timeline for closing that gap, since this PR's own bundle.mlir output is otherwise not yet exercised end-to-end on real hardware?

Verification notes

- Confirmed SymbolKind.dimension/is_dimension already exist on main (added by earlier work), so this PR is purely additive in bundle.py.
- Traced compute_ops.py::generate_sdsc (lines 444–458, 1057–1060): dimension SymbolKinds are registered unconditionally (regardless of use_symbols), while local_symbol_kind (kernel/pool/derived kinds) is only populated when use_symbols=True. This confirms the PR's core assumption — when use_symbols=False, any SymbolKind present must be a dimension kind — is correct, validating both the new AssertionError and the "independent of use_symbols" claim in the docstring.
- Verified _extract_symbol_ids's docstring claim ("dimension IDs have lower-magnitude negatives, so scanning them first keeps ids sorted") — this function operates on one SDSC's JSON at a time (called per-OpSpec in _emit_specs), and within a single SDSC dimension symbol IDs are always allocated before address symbol IDs (compute_ops.py lines 444–458), so the claim holds. (I initially worried this would break across multiple SDSCs sharing a bundle-wide symbol_id_offset counter, but that's moot since the function never sees more than one SDSC's data.)
- DCO: all commits signed off by Nethra Khandige. License headers present and correctly dated 2026 on the new test file.
- CI: all jobs pass, including the new Test Symbolic Dim Bundle job (52s). Test Launch Jobplan and Test Spyre are currently pending on a re-run (they had transiently failed earlier in the run, consistent with the StreamInErrorState-style infra flakiness seen on other PRs this session) — worth a final glance once that re-run completes, but nothing in the diff touches jobplan launch code.

Verdict

APPROVE (pending the two currently-rerunning CI jobs landing green — no code concerns block merge).

File: torch_spyre/_inductor/codegen/bundle.py (hunk 4 above)
Comment (nethra-khandige) [reply torch_spyre/_inductor/codegen/bundle.py:154]: Thanks @Vikram123-byte , you're right that the sdsc files get written before bundle.mlir, so a failure in between would leave a partial dir.
However, this partial-artifact issue where output_dir is left with no bundle.mlir could happen even without symbolic shapes work. For now a failed compile aborts before anything reads the bundle, and the sdsc and bundle files are written in truncate mode so the next run just overwrites them. So nothing consumes a half-written output right now.
As @dgrove-oss  mentioned in his review comment, the exception it points at isn't actually reachable today (the code added here in this PR is more of a defensive check). 
So it feels like a generate_bundle cleanup rather than part of the symbolic-dim work. We'll take it up in a separate issue/PR.

cc: @dgrove-oss @pradghos @Vivek1258

**Review by Vikram123-byte — APPROVED** [review]
LGTM, let @dgrove-oss take final look.

**Review by dgrove-oss — APPROVED** [review]
lgtm

**Comment by dgrove-oss** [comment]
Holding off on enabling merge until Ci is healthy; but intend to merge as soon as we can.
